File: ios/Goo/Inset.swift
import Foundation
import Yaml

struct Inset {
    let left: Double
    let right: Double
    let top: Double
    let bottom: Double
    
    static let zero = Inset()
    
    init(left: Double = 0, right: Double = 0, top: Double = 0, bottom: Double = 0) {
        self.left = left
        self.right = right
        self.top = top
        self.bottom = bottom
    }
    
    init(_ value: Double) {
        left = value
        right = value
        top = value
        bottom = value
    }
    
    init?(_ yaml: Yaml) {
        let left = yaml["left"].double
        let right = yaml["right"].double
        let top = yaml["top"].double
        let bottom = yaml["bottom"].double
        if left != nil || right != nil || top != nil || bottom != nil {
            self.left = left ?? 0
            self.right = right ?? 0
            self.top = top ?? 0
            self.bottom = bottom ?? 0
        } else {
            return nil
        }
    }
    
}

func -(a: CGRect, b: Inset) -> CGRect {
    let left = CGFloat(b.left)
    let right = CGFloat(b.right)
    let top = CGFloat(b.top)
    let bottom = CGFloat(b.bottom)
    return CGRect(
        x: a.origin.x + left,
        y: a.origin.y + top,
        width: a.size.width - left - right,
        height: a.size.height - top - bottom
    )
}

File: android/app/src/main/java/org/haralovich/goo/Inset.kt
package org.haralovich.goo

import com.fasterxml.jackson.databind.JsonNode

class Inset(
    val left: Float = 0f,
    val right: Float = 0f,
    val top: Float = 0f,
    val bottom: Float = 0f
) {

    companion object  {

        val zero = Inset()

        fun parse(json: JsonNode): Inset? {
            val left = json.path("left").floatOption
            val right = json.path("right").floatOption
            val top = json.path("top").floatOption
            val bottom = json.path("bottom").floatOption
            if (left != null || right != null || top != null || bottom != null) {
                return Inset(left ?: 0f, right ?: 0f, top ?: 0f, bottom ?: 0f)

            } else {
                return (
                    Vector.parse(json)?.let { Inset(it) } ?:
                    json.floatOption?.let { Inset(it) }
                )
            }
        }

    }

    constructor(value: Float) : this(value, value, value, value)

    constructor(edge: Vector) : this(edge.x, edge.x, edge.y, edge.y)

    operator fun plus(other: Inset): Inset {
        return Inset(left + other.left, right + other.right, top + other.top, bottom + other.bottom)
    }

    override fun toString(): String {
        return "Inset($left, $top, $right, $bottom)"
    }

}